lib/twitter/rest/upload_utils.rb

```diff
@@ -8,39 +8,53 @@ module UploadUtils
       # Uploads images and videos. Videos require multiple requests and uploads in chunks of 5 Megabytes.
       # The only supported video format is mp4.
       #
-      # @see https://dev.twitter.com/rest/public/uploading-media
+      # @see https://developer.twitter.com/en/docs/media/upload-media/uploading-media/media-best-practices
       def upload(media, media_category_prefix: 'tweet')
         return chunk_upload(media, 'video/mp4', "#{media_category_prefix}_video") if File.extname(media) == '.mp4'
         return chunk_upload(media, 'image/gif', "#{media_category_prefix}_gif") if File.extname(media) == '.gif' && File.size(media) > 5_000_000
 
         Twitter::REST::Request.new(self, :multipart_post, 'https://upload.twitter.com/1.1/media/upload.json', key: :media, file: media).perform
       end
 
-      # rubocop:disable MethodLength
+      # @see https://developer.twitter.com/en/docs/media/upload-media/uploading-media/chunked-media-upload
       def chunk_upload(media, media_type, media_category)
         init = Twitter::REST::Request.new(self, :post, 'https://upload.twitter.com/1.1/media/upload.json',
                                           command: 'INIT',
                                           media_type: media_type,
                                           media_category: media_category,
                                           total_bytes: media.size).perform
+        append_media(media, init[:media_id])
+        media.close
+        finalize_media(init[:media_id])
+      end
 
+      # @see https://developer.twitter.com/en/docs/media/upload-media/api-reference/post-media-upload-append
+      def append_media(media, media_id)
         until media.eof?
           chunk = media.read(5_000_000)
           seg ||= -1
           Twitter::REST::Request.new(self, :multipart_post, 'https://upload.twitter.com/1.1/media/upload.json',
                                      command: 'APPEND',
-                                     media_id: init[:media_id],
+                                     media_id: media_id,
                                      segment_index: seg += 1,
                                      key: :media,
                                      file: StringIO.new(chunk)).perform
         end
+      end
 
-        media.close
-
-        Twitter::REST::Request.new(self, :post, 'https://upload.twitter.com/1.1/media/upload.json',
-                                   command: 'FINALIZE', media_id: init[:media_id]).perform
+      # @see https://developer.twitter.com/en/docs/media/upload-media/api-reference/post-media-upload-finalize
+      # @see https://developer.twitter.com/en/docs/media/upload-media/api-reference/get-media-upload-status
+      def finalize_media(media_id)
+        response = Twitter::REST::Request.new(self, :post, 'https://upload.twitter.com/1.1/media/upload.json',
+                                              command: 'FINALIZE', media_id: media_id).perform
+        3.times do
```

Comment:
Theoretically we should use `loop` but I chose to use `3.times` in order to avoid the risk of infinite loop. Twitter returns the number of seconds we need to wait before checking the status again in `response[:processing_info][:check_after_secs]` so we use `sleep` with this value.

Reply:
IMO we should wrap the `sleep` in a `Timeout::timeout` method so that we don't have a stuck process in case Twitter bugs and gives us an abnormal waiting time.

Reply:
I like the third option :+1:

Reply:
It was my favorite option as well :slightly_smiling_face: 

I implemented it in 20863f8 with a small variation. Instead of wrapping just the `STATUS` part, I wrapped the whole chunk upload.

The timeout can be configured through `Twitter::Client#timeouts[:upload]`.
By default it won't timeout, but if one is provided and reached a `Twitter::Error::TimeoutError` will be raised.